keps/sig-node/5328-node-capabilities/README.md

```diff
@@ -0,0 +1,821 @@
+*   Other API options are explored in the [Alternatives](?tab=t.0#heading=h.vfsx3j6ygclj) section. \
+
+
+### Kube-Scheduler Changes
+
+To enable the scheduler to filter nodes based on advertised capabilities, this KEP proposes the introduction of a new [Filter Plugin](https://kubernetes.io/docs/concepts/scheduling-eviction/scheduling-framework/#filter) - `NodeCapabilityFilter`. This plugin will be part of the default scheduler and will be enabled when the `NodeCapabilities` feature gate is active. Additionally, the filter plugin would skip any capability checks for daemonset pods.
+
+`NodeCapabilityFilter` plugin would 
+
+*   Inspect the PodSpec to infer a set of required NodeCapability key-value pairs. This inference is performed using imperative logic built into the scheduler plugin (without new, explicit capability request fields in the PodSpec).
+*   For each node being evaluated, the plugin will access the `node.status.capabilities` map. It will then iterate through the pod's inferred requirements and check if each required capability is present and matches on the node. If any inferred requirement is not met, the plugin will filter the node, making it unschedulable for the pod.
+*   On a mismatch, the plugin will return an FailedScheduling status with a reason that includes the capability that was not found on Nodes.
+
+
+### Admission Controller Changes
+
+To enable the validation of API requests against Node Capabilities, this KEP proposes the introduction of a new admission controller plugin `NodeCapabilityValidator `that will be enabled when the NodeCapabilities feature gate is active. This is necessary because the validation logic that requires cross-referencing objects i,e looking up the `Node` object that a `Pod` is bound to (spec.nodeName). THe admission controller would 
+
+*   Inspects incoming pod update request and infer if it depends on any node capability. 
+*   It verifies that `pod.spec.nodeName` is set. If the pod is not yet bound to a node, the validator takes no action.
+*   Retrieve the `Node` object corresponding to` pod.spec.nodeName` and check the node.Status.capabilities map of the fetched Node object.
+*   If the capability is missing or has a non-matching value, the request is marked as invalid
+
+### Example Walkthrough
+
+**New Features and Version Skew Management**
+
+1. When developing a new Kubelet feature Eg:[ In-Place Pod Resize for Guaranteed QoS pods](https://github.com/kubernetes/enhancements/issues/5294), the skew management strategy involves
+    *   Modifying the Kubelet to advertise the feature as a Node Capability if the feature gate `InPlacePodVerticalScalingExclusiveCPUs` is active. Kubelet would add <code>[kubernetes.io/feature/inPlacePodResizeExclusiveCpus](http://kubernetes.io/feature/inPlacePodResizeExclusiveCpus): true </code>in<code> status.capabilities</code>.
```

Comment:
This is another example of a feature that depends on both Kubelet configuration & feature gate. Would kubelet publish both the feature gate capability, and the CPU policy as a capability?